hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/sps/BlockStorageMovementNeeded.java

```diff
@@ -248,13 +251,22 @@ public void run() {
                   pendingWorkForDirectory.get(startINode);
               if (dirPendingWorkInfo != null
                   && dirPendingWorkInfo.isDirWorkDone()) {
-                ctxt.removeSPSHint(startINode);
+                try {
+                  ctxt.removeSPSHint(startINode);
+                } catch (FileNotFoundException e) {
+                  // ignore if the file doesn't already exist
+                  startINode = null;
+                }
                 pendingWorkForDirectory.remove(startINode);
               }
             }
             startINode = null; // Current inode successfully scanned.
           }
         } catch (Throwable t) {
+          retryCount++;
+          if (retryCount >= 3) {
+            startINode = null;
+          }
```

Comment:
@liubingxing 
- Let's define the constant of the max retry count (`private static final int MAX_RETRY_COUNT = 3;`) in `SPSPathIdProcessor`.
- How about logging a message when skipping the inode?
```suggestion
          retryCount++;
          if (retryCount >= MAX_RETRY_COUNT) {
            LOG.warn("Skipping this inode {} due to too many retries.", startINode);
            startINode = null;
          }
```

- And I think it's better to move the retry logic to the end of the catch block.

Reply:
@tasanuma Thanks for your advices. I updated the code.